### ISSUE TITLE
Project proposal: ASGI profiling middleware using yappi

### ISSUE THREAD

**Issue by dmontagu** [florimondmanca/awesome-asgi#17]
(Apologies in advance if this isn't an appropriate place to start a discussion like this!)

-------

It looks like yappi support for coroutine profiling is just about complete: https://github.com/tiangolo/fastapi/issues/701#issuecomment-558503062

This should enable ASGI-compatible request-specific profiling (during development at least; yappi probably adds too much overhead to make sense during production). In particular, I think this could make it relatively easy to perform performance regression testing, etc.

If you look in that thread, you'll also see it can be used to display profiling info directly inside the browser. 🔥

-----

I think it would be great if we could throw together a small ASGI middleware profiling package similar to the proposal from [this comment](https://github.com/tiangolo/fastapi/issues/701#issuecomment-552547543), making use of the new yappi capabilities once they are released.

Eventually, I think it might make sense to try to package up functionality similar to what you'd expect from `pytest --profile-svg` in a non-asyncio context, as described in [this comment](https://github.com/encode/starlette/issues/686#issue-511251521).

-----

If anyone is interested in collaborating on this let me know! I don't have experience publishing packages on PyPI so it would be great if someone more knowledgeable about that process could help out.

**Comment by florimondmanca** [comment]
Hi! This sounds very exciting. I think there’s some tinkering and research necessary to fit the various use cases, and perhaps make the middleware more generic (read: lower-level implementation) while addressing those use cases effectively.

FWIW, timing-asgi, listed in here, is a much simpler attempt at providing timing for ASGI apps. Yappi seems to be on an entirely different level of power though. :)

I’d definitely be interested in seeing what comes out of this! Feel free to ping me anywhere you think if relevant, if I can help with a more holistic view of the ASGI ecosystem.

**Comment by sm-Fifteen** [comment]
> Hi! This sounds very exciting. I think there’s some tinkering and research necessary to fit the various use cases, and perhaps make the middleware more generic (read: lower-level implementation) while addressing those use cases effectively.
> 
> FWIW, timing-asgi, listed in here, is a much simpler attempt at providing timing for ASGI apps. Yappi seems to be on an entirely different level of power though. :)

It probably wouldn't be difficult to clean it up (I thought starting and stopping Yappi for each request was the right thing to do at the time, for instance, but it turns out you're supposed to start it once and leave it running since the profiler is global) turn this Starlette Middleware into a generic ASGI middleware, I just wrote it that way to keep the code example short and easy to follow. `timing-asgi` looks like a simple implementation that probably does a good job at what it tries to do, but the performance breakdown per request was an important factor in why I started looking into actual profiling solutions (since FastAPI can run into performance bottlenecks for things like JSON encoding and serializing, and you need to be able to tell those appart from slow database queries and the like).

> If anyone is interested in collaborating on this let me know! I don't have experience publishing packages on PyPI so it would be great if someone more knowledgeable about that process could help out.

I doubt there's going to be a whole lot of work left to do on this once Yappi support for coroutines lands, but I'd be willing to help you on this, sure!

**Comment by euri10** [comment]
as I will definitely use it I'm game to help in any kind of way, that said you guys are way better suited to code the stuff ! I'll still be very happy to try break it as a user :rabbit: already begun, loving it

**Comment by florimondmanca** [comment]
Just stopping by to refer to the associated issue on the `yappi` repo: https://github.com/sumerc/yappi/issues/21.

I'll close this since I don't see any immediate action we can take relative to *this* repo, but again the idea sure sounds like fun and interesting. I subscribed to the issue above! If anything gets out there y'all may want to post something over https://discuss.encode.io. :-)

Thanks!

**Comment by sumerc** [comment]
Just discovered this thread by your reference. I just wanted to inform: the code in sumerc/yappi#21 is actually merged to master and released today.

Cheers!

**Comment by sm-Fifteen** [comment]
@florimondmanca: Would something like this fit the standards for this repo? There's still a few things  I need to figure out, so I haven't created a repo for this yet, but I think it's shaping up nicely.

```py
from contextvars import ContextVar
from typing import Dict
import re
from ctypes import c_long

import yappi
from yappi import YFuncStats

_yappi_request_id: ContextVar[int] = ContextVar('_yappi_request_id', default=-1)

def _get_context_id() -> int:
    return _yappi_request_id.get()

class ServerTimingMiddleware:
    """Timing middleware for ASGI HTTP applications

    The resulting profiler data will be returned through the standard
    `Server-Timing` header for all requests.

    Args:
        app (ASGI v3 callable): An ASGI application

        calls_to_track (Dict[str,str]): A dict of qualified function names
            keyed by desired output metric name.
            
            Metric names must consist of a single rfc7230 token
            and qualified function names can be obtained with
            `my_function.__qualname__` for both functions and coroutines.

    .. _Server-Timing sepcification:
        https://w3c.github.io/server-timing/#the-server-timing-header-field
    """

    def __init__(self, app, calls_to_track: Dict[str, str]):
        for metric_name in calls_to_track.keys():
            # https://httpwg.org/specs/rfc7230.html#rule.token.separators
            # USASCII (7 bits), only visible characters (no non-printables or space), no double-quote or delimiter
            if not metric_name.isascii() or not metric_name.isprintable() or re.search(r'[ "(),/:;<=>?@\[\\\]{}]', metric_name) is not None:
                # TODO: length == 0
                raise ValueError('"{}" contains an invalid character for a Server-Timing metric name'.format(metric_name))

        self.app = app
        self.calls_to_track = calls_to_track

        yappi.set_context_id_callback(_get_context_id)
        yappi.set_clock_type("wall")

        # TODO: Schedule some kind of periodic profiler cleanup with clear_stats()
        yappi.start()

    async def __call__(self, scope, receive, send):
        ctx_id = id(scope)
        # Longs on Windows are only 32 bits, but memory adresses on 64 bit python are 64 bits
        ctx_id = abs(c_long(ctx_id).value) # Ensure it fits inside a long, truncating if necessary
        _yappi_request_id.set(ctx_id)

        def wrapped_send(response):
            if response['type'] == 'http.response.start':
                tracked_stats: Dict[str, YFuncStats] = {
                    name: yappi.get_func_stats({"name": call_to_track, "ctx_id": ctx_id})
                    for name, call_to_track in self.calls_to_track.items()
                }

                server_timing = [
                    f"{name};dur={(stats.pop().ttot * 1000):.3f}".encode('ascii')
                    for name, stats in tracked_stats.items()
                    if not stats.empty()
                ]

                if server_timing:
                    # FIXME: Doesn't check if a server-timing header is already set
                    response['headers'].append(["server-timing", b','.join(server_timing)])

            return send(response)

        await self.app(scope, receive, wrapped_send)
```

Usage looks something like this (I'm using FastAPI to demonstrate, but this is a generic ASGI middleware):

```py
app = FastAPI()

@app.get("/")
async def context_id_endpoint():
    await asyncio.sleep(0.5)
    return _get_context_id()


track: Dict[str, str] = {
    "db_exec": sqlalchemy.engine.base.Engine.execute.__qualname__,
    "pydantic": fastapi.routing.serialize_response.__qualname__,
    "render": Response.render.__qualname__,
    "endpoint": context_id_endpoint.__qualname__
}
app.add_middleware(ServerTimingMiddleware, calls_to_track=track)
```

A few quirks I've yet to address:
* The int-to-long casting to workaround sumerc/yappi#38 on Windows.
* No check for existing Server-Timing headers before appending one to the list, so it may get set multiple times if something else has already set that header (which is technically incorrect).
* The profiler stats are never cleared, which means the server will leak if left running for extended periods of time. Yappi currently offers no way to incrementally clear collected data, so profiler data has to be cleared completely (in a way that's schedule independently from the rest of the application), which would affect all currently running tasks. (@sumerc: Any advice on this?)

**Comment by sumerc** [comment]
> The int-to-long casting to workaround sumerc/yappi#38 on Windows.

I was going to message you today regarding sumerc/yappi#38. I think it is fixed(branch fix-issue38). Please you also verify on your side and let's continue discussion from the issue.

> Yappi currently offers no way to incrementally clear collected data, so profiler data has to be cleared completely (in a way that's schedule independently from the rest of the application), which would affect all currently running tasks.

clear_stats() with filtering can be done as we have discussed before. If this is that urgent, let's open an issue and follow up there. No problem on my side.

One thing I noticed while looking at the code, it seems you are using `set_context_id` instead of `set_tag_callback`, are you sure about that? The last version of the code we have tested is not like this and if you use `set_context_id` you will overwrite the internal `thread->ctx` mapping which might cause problems for same coroutines running on different threads? Here is the version of the code I am referring: https://github.com/tiangolo/fastapi/issues/701#issuecomment-558568564

**Comment by sm-Fifteen** [comment]
> > The int-to-long casting to workaround [sumerc/yappi#38](https://github.com/sumerc/yappi/issues/38) on Windows.
> 
> I was going to message you today regarding [sumerc/yappi#38](https://github.com/sumerc/yappi/issues/38). I think it is fixed(branch fix-issue38). Please you also verify on your side and let's continue discussion from the issue.

I noticed you'd been working on this, yes, thank you. I was just mentionning it here because I left the workaround in the code for the time being. You're right, though, we should probably keep the discussion on that topic inside its own issue.

> > Yappi currently offers no way to incrementally clear collected data, so profiler data has to be cleared completely (in a way that's schedule independently from the rest of the application), which would affect all currently running tasks.
> 
> clear_stats() with filtering can be done as we have discussed before. If this is that urgent, let's open an issue and follow up there. No problem on my side.

Right, I couldn't find the exact message where we'd talked about this. I wouldn't call it "urgent" *per se*, but I'll admit it's slightly inconveinient for this use case, since it means the middlware can't just be activated and then forgotten about. I'll probably open a yappi issue about it later so that can be discussed in more details.

> One thing I noticed while looking at the code, it seems you are using `set_context_id` instead of `set_tag_callback`, are you sure about that? The last version of the code we have tested is not like this and if you use `set_context_id` you will overwrite the internal `thread->ctx` mapping which might cause problems for same coroutines running on different threads? Here is the version of the code I am referring: [tiangolo/fastapi#701 (comment)](https://github.com/tiangolo/fastapi/issues/701#issuecomment-558568564)

Ah, good point, I hadn't noticed the change from `yappi.set_context_id_callback(get_context_id)` to `yappi.set_tag_callback(get_context_id)` (I had actually forgotten about that new API). I'll be sure to use that instead; thank you for pointing it out.
